torchvision/ops/boxes.py

```diff
@@ -382,7 +382,39 @@ def _box_diou_iou(boxes1: Tensor, boxes2: Tensor, eps: float = 1e-7) -> Tuple[Te
     # distance between boxes' centers squared.
     return iou - (centers_distance_squared / diagonal_distance_squared), iou
 
+def masks_to_boundaries(masks: torch.Tensor, dilation_ratio: float = 0.02) -> torch.Tensor:
+    """
+    Compute the boundaries around the provided masks using morphological operations.
+
+    Returns a tensor of the same shape as the input masks containing the boundaries of each mask.
+
+    Args:
+    masks (Tensor[N, H, W]): masks to transform where N is the number of masks
+    and (H, W) are the spatial dimensions.
+    dilation_ratio (float, optional): ratio used for the dilation operation. Default: 0.02
+
+    Returns:
+    Tensor[N, H, W]: boundaries
+    """
+    # If no masks are provided, return an empty tensor
+    if masks.numel() == 0:
+        return torch.zeros_like(masks)
+
+    n, h, w = masks.shape
+    img_diag = math.sqrt(h ** 2 + w ** 2)
+    dilation = int(round(dilation_ratio * img_diag))
+    selem_size = dilation * 2 + 1
+    selem = torch.ones((n, 1, selem_size, selem_size), device=masks.device)
 
+    # Compute the boundaries for each mask
+    masks = masks.float().unsqueeze(1)
+    eroded_masks = F.conv2d(masks, selem, padding=dilation, groups=n)
+    eroded_masks = (eroded_masks == selem.view(n, -1).sum(1, keepdim=True)).byte()  # Make the output binary
+
+    contours = masks.byte() - eroded_masks
+
+    return contours.squeeze(1)
```

Comment:
I do not think this code works as expected. Here is my test example and it fails in multiple places:
```python
import torch
import numpy as np
from PIL import ImageDraw, Image

mask = torch.zeros(4, 32, 32, dtype=torch.bool)
mask[0, 1:10, 1:10] = True
mask[0, 12:20, 12:20] = True
mask[0, 15:18, 20:32] = True

mask[1, 15:23, 15:23] = True
mask[1, 22:33, 22:33] = True

mask[2, 1:5, 22:30] = True
mask[2, 5:14, 25:27] = True


pil_img = Image.new("L", (32, 32))

draw = ImageDraw.Draw(pil_img)
draw.ellipse([2, 7, 26, 26], fill=1, outline=1, width=1)

mask[3, ...] = torch.from_numpy(np.asarray(pil_img))


import math
from torch.nn import functional as F

dilation_ratio = 0.05
masks = mask.clone()

n, h, w = masks.shape
img_diag = math.sqrt(h ** 2 + w ** 2)
dilation = int(round(dilation_ratio * img_diag))
selem_size = dilation * 2 + 1
selem = torch.ones((n, 1, selem_size, selem_size), device=masks.device)


# Compute the boundaries for each mask
masks = masks.float().unsqueeze(1)
eroded_masks = F.conv2d(masks, selem, padding=dilation, groups=n)
eroded_masks = (eroded_masks == selem.view(n, -1).sum(1, keepdim=True)).byte()  # Make the output binary

contours = masks.byte() - eroded_masks
contours. = contours.squeeze(1)
```

Error: 
```
---> 17 eroded_masks = (eroded_masks == selem.view(n, -1).sum(1, keepdim=True)).byte()  # Make the output binary

RuntimeError: The size of tensor a (32) must match the size of tensor b (4) at non-singleton dimension 2
```

Masks:
![image](https://github.com/pytorch/vision/assets/2459423/a2786409-5b98-46dc-ae40-0befadfdcfe7)


Error is related to `masks = masks.float().unsqueeze(1)` where we may need to unsqueeze(0) instead.
But if fixed like that, the next line does not make much sense IMO:
```
eroded_masks = (eroded_masks == selem.view(n, -1).sum(1, keepdim=True)).byte()
```
as eroded_masks shape wont match the size of conv weights...

Sorry, if I'm missing something...

Reply:
What do you think about:
```python
import torch
import numpy as np
from PIL import ImageDraw, Image
import math
from torch.nn import functional as F
import matplotlib.pyplot as plt

mask = torch.zeros(4, 32, 32, dtype=torch.bool)
mask[0, 1:10, 1:10] = True
mask[0, 12:20, 12:20] = True
mask[0, 15:18, 20:32] = True

mask[1, 15:23, 15:23] = True
mask[1, 22:33, 22:33] = True

mask[2, 1:5, 22:30] = True
mask[2, 5:14, 25:27] = True

pil_img = Image.new("L", (32, 32))
draw = ImageDraw.Draw(pil_img)
draw.ellipse([2, 7, 26, 26], fill=1, outline=1, width=1)
mask[3, ...] = torch.from_numpy(np.asarray(pil_img))

dilation_ratio = 0.05
masks = mask.clone()

n, h, w = masks.shape
img_diag = math.sqrt(h ** 2 + w ** 2)
dilation = int(round(dilation_ratio * img_diag))
selem_size = dilation * 2 + 1
selem = torch.ones((1, 1, selem_size, selem_size), device=masks.device)

# Compute the boundaries for each mask
masks = masks.float().unsqueeze(1)
eroded_masks = torch.zeros_like(masks)

#for i in range(n):
#    eroded_masks[i] = F.conv2d(masks[i].unsqueeze(0), selem, padding=dilation)
eroded_masks = F.conv2d(masks, selem, padding=dilation)

eroded_masks = (eroded_masks == selem.view(-1).sum()).byte()  # Make the output binary
contours = masks.byte() - eroded_masks
contours = contours.squeeze(1)

# Visualize the results
fig, ax = plt.subplots(n, 3, figsize=(10, 10))

for i in range(n):
    ax[i, 0].imshow(mask[i], cmap='gray')
    ax[i, 1].imshow(eroded_masks[i].squeeze(), cmap='gray')
    ax[i, 2].imshow(contours[i], cmap='gray')

plt.show()
```
![immagine](https://github.com/pytorch/vision/assets/1710528/e5d42500-a778-4248-9ec8-1b47d19ef8ff)

Reply:
@bhack why do we need dilation_ratio ? I think we can do the following without extra parametrization:
```python
masks = masks.float().unsqueeze(1)
w_size = 3
w = torch.ones((1, 1, w_size, w_size), device=masks.device) / (w_size ** 2)
eroded_masks = F.conv2d(masks, w, padding=1)
contours = (masks - eroded_masks) > 0
contours = contours.squeeze(1)
```
what do you think ?

Reply:
It is in the paper official implementation

https://github.com/bowenc0221/boundary-iou-api/blob/master/boundary_iou/utils/boundary_utils.py#L12

But also in the more classical F score (Davis dataset/challenge official eval kit).

https://github.com/davisvideochallenge/davis2017-evaluation/blob/master/davis2017/metrics.py#L57

As this is often a preprocessing step used in the boundary overlapping metrics (BoundaryIOU/Boundary F-Score) the dilate will give the control over the tolerance of the exact boundaries overlapping of the boundaries.

In both the papers they talked about bipartite graph matching but then they have always  approximated with morphological ops.

Of you see the F/Davis  case impl there is also an option where the tolerance/dilate Is defined by the input resolution.

Reply:
Thanks for the links. According to https://github.com/davisvideochallenge/davis2017-evaluation/blob/master/davis2017/metrics.py#L57 code, mask to boundary is done without using any parameters, see `_seg2bmap`:
https://github.com/davisvideochallenge/davis2017-evaluation/blob/ac7c43fca936f9722837b7fbd337d284ba37004b/davis2017/metrics.py#L122
Anyway, I see why they have `dilation_ratio` arg.


However, previously I missed the issue description and the context for this PR:

> A mask to boundary API is useful for implementing many segmentation metrics used in many dataset and challenges (Davis F score, BoundaryIOU, etc..).
It could be also used more generally for visualization tasks.

In this case, I'm not very sure about torchvision's interest in following line by line what does https://github.com/bowenc0221/boundary-iou-api as 1) IMO we wont be able to reproduce `cv2.erode` behaviour and 2) as such helper function can be used within a metric implementation, it should be carefully tested vs ref implementation in a lot of corner cases etc (and this is not the role of torchvision, IMO).

In general, a method to produce mask to edges (sort of edge detector) could make sense like mask to bboxes.

Reply:
> Thanks for the links. According to https://github.com/davisvideochallenge/davis2017-evaluation/blob/master/davis2017/metrics.py#L57 code, mask to boundary is done without using any parameters, see _seg2bmap:
https://github.com/davisvideochallenge/davis2017-evaluation/blob/ac7c43fca936f9722837b7fbd337d284ba37004b/davis2017/metrics.py#L122

Yes but cause in F they are dilating in an extra post-processing step in the metric instead of the `BoundariesIOU` approach (see dilate disk param)
https://github.com/davisvideochallenge/davis2017-evaluation/blob/master/davis2017/metrics.py#L77

> In this case, I'm not very sure about torchvision's interest in following line by line what does https://github.com/bowenc0221/boundary-iou-api as 1) IMO we wont be able to reproduce cv2.erode behaviour and 2) as such helper function can be used within a metric implementation, it should be carefully tested vs ref implementation in a lot of corner cases etc (and this is not the role of torchvision, IMO).

I've tested another early implementation with some inputs but the Boundary IOU paper reference impl doesn't have a test suite.

> In general, a method to produce mask to edges (sort of edge detector) could make sense like mask to bboxes.

Let me know as I am mainly interested to [achieve the metric](https://github.com/Lightning-AI/torchmetrics/issues/1500)  and eventually to contribute also an intermediate function here in the case it could be compatible and useful for other contexts/domain.

Reply:
Also as I see you are also a member of the MONAI project I see that you have already something similar it but still rely on a non-Pytorch implementation:
https://github.com/Project-MONAI/MetricsReloaded/blob/main/MetricsReloaded/metrics/pairwise_measures.py#L963